src/librustc_codegen_llvm/debuginfo/mod.rs

```diff
@@ -59,14 +59,19 @@ const DW_TAG_arg_variable: c_uint = 0x101;
 
 /// A context object for maintaining all state needed by the debuginfo module.
 pub struct CrateDebugContext<'a, 'tcx> {
+    // FIXME(eddyb) we need this?! at all?!
```

Comment:
A const int needs to be creates when emitting enum debuginfo. With my WIP PR the CodegenCx is no longer passed to all debuginfo functions, so the context is necessary to be stored here. I don't know where it is currently needed though.

Reply:
> `CodegenCx` is no longer passed to all debuginfo functions

I don't see how that can work, it should still involve `CodegenCx` one way or another, and most of this debuginfo state should live in it.

Reply:
Most of the things required for debuginfo live in TyCtxt, at least what is needed for the typeinfo. In my PR I added a tcx field.

Reply:
That doesn't really work for LLVM though... I think `CrateDebugContext` is a misleading name and it should be called `DebugData` or `DebugBuilder`.

The only way this would make sense to me is if there was an entirely different `rustc_codegen_debuginfo` crate with abstractions separate from `rustc_codegen_ssa` (but which the latter used, because it's needed for using locations and variables in a function).

But even then... they still feel too coupled.

Reply:
`DebugBuilder` would be `llvm::DIBuilder`. `CrateDebugContext` is in my opinion everything non function local that is needed to create debuginfo. You ask it for the debuginfo of a `Ty<'tcx>`, and it will ask the `DIBuilder` to create the corresponding debuginfo. Think about it in terms of the split between `cg_ssa::FunctionCx` and `cg_llvm::Builder`. The former is like `CrateDebugContext`, while the later is like `DIBuilder`. `DIBuilder` is for low-level creation of debuginfo by specifying the name, size, fields (type+offset), .... `CrateDebugContext` is for high-level creation of debuginfo by specifying `Ty<'tcx>`.

Reply:
Okay I was a bit confused about `DIBuilder` so I took another look at it and I can kind of see what you mean.

I would definitely want to rename `CrateDebugContext` to `DIBuilder` or `LlvmDbgBuilder` or something like that if we want it to be disjoint from `CodegenCx`, I'm still worrying a bit about duplication I guess.